quartz/components/renderPage.tsx

```diff
@@ -52,33 +52,25 @@ export function pageResources(
   }
 }
 
-let pageIndex: Map<FullSlug, QuartzPluginData> | undefined = undefined
-function getOrComputeFileIndex(allFiles: QuartzPluginData[]): Map<FullSlug, QuartzPluginData> {
-  if (!pageIndex) {
-    pageIndex = new Map()
-    for (const file of allFiles) {
-      pageIndex.set(file.slug!, file)
-    }
-  }
-
-  return pageIndex
-}
-
 export function renderPage(
   cfg: GlobalConfiguration,
   slug: FullSlug,
   componentData: QuartzComponentProps,
   components: RenderComponents,
   pageResources: StaticResources,
 ): string {
+  // make a deep copy of the tree so we don't remove the transclusion references
+  // for the file cached in contentMap in build.ts
+  const root = structuredClone(componentData.tree) as Root
+
   // process transcludes in componentData
-  visit(componentData.tree as Root, "element", (node, _index, _parent) => {
+  visit(root, "element", (node, _index, _parent) => {
     if (node.tagName === "blockquote") {
       const classNames = (node.properties?.className ?? []) as string[]
       if (classNames.includes("transclude")) {
         const inner = node.children[0] as Element
         const transcludeTarget = inner.properties["data-slug"] as FullSlug
-        const page = getOrComputeFileIndex(componentData.allFiles).get(transcludeTarget)
+        const page = componentData.allFiles.find((f) => f.slug === transcludeTarget)
```

Comment:
this is still really heavy on large sites with lots of transcludes, can we keep the memoization but within `renderPage`?

Reply:
This should be lightweight because only the necessary files (leaf node ancestors) are provided in `componentData.allFiles` in fast rebuild mode. So if the transclusion graph looks like a -> b -> c -> c.html, only a, b, c are passed in.

In normal rebuild mode we'll get incorrect behaviour with memoization in case the file's transclusion has changed as part of the edit. Also I think linear search through 1000s of files should still be quick?